### DISCUSSION TITLE
Basic Text Editor / IDE Functionality

### DISCUSSION THREAD

**Discussion by tinkerctl** [fyne-io/fyne#5125] [General]
Is there a better parent issue / discussion than [this feature request](https://github.com/fyne-io/fyne/issues/2618) for supporting basic text editor functionality?

From what I can tell based on browsing the API, and sample projects such as [defyne](https://github.com/fyne-io/defyne) and [notes](https://github.com/fynelabs/notes), the only method for editing large text files is the basic Entry type, which seems to lack support for cursor movement, highlighting, and some other primitives that may allow for tool tipping / intellisene like behavior.

I'm new to this project and would like any pointers on existing design decisions for this feature so that I can determine if it would be feasible for me to begin work on a PR, potentially to make TextGrid editable.

**Comment by andydotxyz** [comment]
I don’t understand what you mean about Entry widget lacking support for cursor movement and selection - it supports all of that…

You might have to be more specific about what you are trying to do but can’t.

**Reply by tinkerctl** [reply]
Apologies. 

For context, I'm looking to create a super simple auto-completion like you'd see in any IDE, for one specific pattern. I could probably make this work with a select widget at the bottom of the screen, set the focus and cursor back to where it belongs once the selection is made. It'd be nice if this was directly above the text instead of anchored somewhere.

As for functionality of the Entry widget... I did get the cursor position working. I'm still not sure where tooltips/hovering support would be, though.

At a minimum, I'm eventually looking to take advantage of an editable `TextGrid` for easier reading / manipulation of each line. Is there a plan for this? If so, some pointers on existing convos or rough idea on how this would be implemented would allow me to begin researching whether I can try to contribute.

**Reply by andydotxyz** [reply]
>  I'm still not sure where tooltips/hovering support would be, though.

There is no built-in support for tooltips. You can use multiple layers in your app/widget to handle overlays or use the PopUp widget API.

> At a minimum, I'm eventually looking to take advantage of an editable TextGrid for easier reading / manipulation of each line. Is there a plan for this? If so, some pointers on existing convos or rough idea on how this would be implemented would allow me to begin researching whether I can try to contribute.

Yes. First we need to add scrolling, then we ned to add real performance like the collection widgets. After that we could overlay edibility.

**Comment by tinkerctl** [comment]
Just for my own reference later, https://github.com/fyne-io/fyne/discussions/5135 brought a proposed RichTextEntry to my attention as well.
